### DISCUSSION TITLE
differences between scipy.signal.welch and pwelch

### DISCUSSION THREAD

**Discussion by bnicenboim** [gjmvanboxtel/gsignal#5] [Q&A] [answered]
Hi, I thought I use this channel (rather than email) in case this is useful to someone else as well.

I'm following this example:https://raphaelvallat.com/bandpower.html trying to use gsignal rather than python's scipy.signal

In the blog post they do:

```
from scipy import signal

# Define window length (4 seconds)
win = 4 * sf
freqs, psd = signal.welch(data, sf, nperseg=win)
```
(with `sf = 100`), and I'm trying to figure out how I would do it with gsignal.
I tried

```
gsignal::pwelch(data, window = win, fs = 100)
```
but I get the PSD at  257 frequencies, while with scipy.signal.welch I get it at 201 frequencies. (BTW, it might be useful to throw a warning when nfft is ignored).

Do you happen to know what's the difference between the implementations?

Thanks!
Best,
Bruno

**Comment by gjmvanboxtel** [comment, marked as answer]
Hi Bruno,

The difference arises because gsignal::pwelch silently pads the input data to the nearest power of 2. This was done to increase processing speed frequency resolution. It mimicked the 'sloppy' parameter in the Octave package signal. At the time, I thought it was a good idea to implement this as a standard, so in your example you had 400 seconds, divided by 2 = 200, and padding to nearest higher power of 2 = 256 + 1 = 257 points.

But come to think of it, it might not have been the best idea (because it is, well, sloppy), so I have removed it in the version just uploaded to Github. Your example now returns 201 points as you would expect. It just did not make it into v0.3-4 which was pushed to CRAN yesterday...

Best,
Geert
